stdlib/public/core/Diffing.swift

```diff
@@ -0,0 +1,782 @@
+        endpoints.storage[i] = (x, y)
+
+        isInsertion = false
+        k -= 1
+      }
+    }
+
+    if pathStart != endpoints.storage[i] {
+      i -= 1; endpoints.storage[i] = pathStart
+    }
+
+    let pathStorage = endpoints.storage
+    endpoints.storage = []
+    return CollectionChanges(pathStorage: pathStorage, pathStartIndex: i)
+  }
+}
+
+/// An index that counts its offset from the start of its collection.
+struct CountingIndex<Base : Comparable> : Equatable {
+  /// The position in the underlying collection.
+  let base: Base
+  /// The offset from the start index of the collection or `nil` if `self` is
+  /// the end index.
+  let offset: Int?
+}
+
+extension CountingIndex : Comparable {
+  static func <(lhs: CountingIndex, rhs: CountingIndex) -> Bool {
+    return (lhs.base, lhs.offset ?? Int.max) < (rhs.base, rhs.offset ?? Int.max)
+  }
+}
+
+/// A collection that counts the offset of its indices from its start index.
+///
+/// You can use `CountingIndexCollection` with algorithms on `Collection` to
+/// calculate offsets of significance:
+///
+///   if let i = CountingIndexCollection("Café").index(of: "f") {
+///     print(i.offset)
+///   }
+///   // Prints "2"
+///
+/// - Note: The offset of `endIndex` is `nil`
+struct CountingIndexCollection<Base : BidirectionalCollection> {
+  let base: Base
+
+  init(_ base: Base) {
+    self.base = base
+  }
+}
+
+extension CountingIndexCollection : BidirectionalCollection {
+  typealias Index = CountingIndex<Base.Index>
+  typealias Element = Base.Element
+
+  var startIndex: Index {
+    return Index(base: base.startIndex, offset: base.isEmpty ? nil : 0)
+  }
+
+  var endIndex: Index {
+    return Index(base: base.endIndex, offset: nil)
+  }
+
+  func index(after i: Index) -> Index {
+    let next = base.index(after: i.base)
+    return Index(
+      base: next, offset: next == base.endIndex ? nil : i.offset! + 1)
+  }
+
+  func index(before i: Index) -> Index {
+    let prev = base.index(before: i.base)
+    return Index(
+      base: prev, offset: prev == base.endIndex ? nil : i.offset! + 1)
+  }
+
+  subscript(position: Index) -> Element {
+    return base[position.base]
+  }
+}
+
+/// Returns the nth [triangular number].
+///
+/// [triangular number]: https://en.wikipedia.org/wiki/Triangular_number
+fileprivate func triangularNumber(_ n: Int) -> Int {
+  return n * (n + 1) / 2
+}
+
+/// A square matrix that only provides subscript access to elements on, or
+/// below, the main diagonal.
+///
+/// A [lower triangular matrix] can be dynamically grown:
+///
+///   var m = LowerTriangularMatrix<Int>()
+///   m.appendRow(repeating: 1)
+///   m.appendRow(repeating: 2)
+///   m.appendRow(repeating: 3)
+///
+///   assert(Array(m.rowMajorOrder) == [
+///     1,
+///     2, 2,
+///     3, 3, 3,
+///   ])
+///
+/// [lower triangular matrix]: http://en.wikipedia.org/wiki/Triangular_matrix
+fileprivate struct LowerTriangularMatrix<Element> {
```

Comment:
```suggestion
fileprivate struct _LowerTriangularMatrix<Element> {
```

Reply:
We also need to add explicit `private`/`fileprivate` modifiers for all members, including the ones defined in the extensions below. (But we don't need more underscores.)